aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -166,259 +224,333 @@ module aptos_framework::fungible_asset {
-    ) acquires FungibleAssetWallet {
-        let FungibleAssetWallet {
-            metadata: _,
-            balance: _,
-            allow_ungated_transfer: _,
-            delete_ref
-        } = move_from<FungibleAssetWallet>(object::object_address(&wallet));
-        object::delete(delete_ref);
+    /// Return `amount` of a given fungible asset.
+    public fun asset_metadata(fa: &FungibleAsset): Object<FungibleAssetMetadata> {
+        fa.metadata
     }
 
-    /// Get the balance of `wallet`.
-    public fun balance(wallet: &Object<FungibleAssetWallet>): u64 acquires FungibleAssetWallet {
-        borrow_fungible_asset(wallet).balance
+    /// Get the underlying metadata object from `MintRef`.
+    public fun mint_ref_metadata(ref: &MintRef): Object<FungibleAssetMetadata> {
+        ref.metadata
     }
 
+    /// Get the underlying metadata object from `TransferRef`.
+    public fun transfer_ref_metadata(ref: &TransferRef): Object<FungibleAssetMetadata> {
+        ref.metadata
+    }
 
-    /// Return `allowed_ungated_transfer`.
-    public fun ungated_transfer_allowed(wallet: &Object<FungibleAssetWallet>): bool acquires FungibleAssetWallet {
-        borrow_fungible_asset(wallet).allow_ungated_transfer
+    /// Get the underlying metadata object from `BurnRef`.
+    public fun burn_ref_metadata(ref: &BurnRef): Object<FungibleAssetMetadata> {
+        ref.metadata
+    }
+
+    /// Transfer `amount` of fungible metadata of `metadata` to `receiver`.
+    /// Note: it does not move the underlying object.
+    public entry fun transfer<T: key>(
+        sender: &signer,
+        metadata: Object<T>,
+        amount: u64,
+        recipient: address,
+    ) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let fa = withdraw(sender, metadata, amount);
+        deposit(recipient, fa);
+    }
+
+    /// Withdraw `amount` of fungible asset from an account's wallet.
+    public fun withdraw<T: key>(
+        owner: &signer,
+        metadata: Object<T>,
+        amount: u64,
+    ): FungibleAsset acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let owner_addr = signer::address_of(owner);
+        assert!(wallet_exists(owner_addr, metadata), error::invalid_argument(EINSUFFICIENT_BALANCE));
+        assert!(
+            ungated_transfer_allowed(owner_addr, metadata),
+            error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED),
+        );
+        let wallet = create_wallet_address_from_account(owner_addr, metadata);
+        withdraw_internal(wallet, amount)
+    }
+
+    /// Deposit fungible assets into an account.
+    public fun deposit(owner: address, fa: FungibleAsset) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        assert!(
+            !wallet_exists(owner, fa.metadata) ||
+                ungated_transfer_allowed(owner, fa.metadata),
+            error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED),
+        );
+        deposit_internal(owner, fa);
```

Comment:
`allow_ungated_transfer` should apply to both withdraw and deposit. you don't want to allow money inflow to a frozen account

Reply:
It does. Both deposit and withdraw check ungated. What's wrong with the current code?